### PR TITLE
Per-User LRU sorting of Datasets

### LINKED ISSUE #2917: Consider user-specific metric when sorting datasets (view count, LRU)
To improve the ranking during filtering datasets even further, we can try to integrate another metric into the ranking algorithm. At the moment, we are using the dice-coefficient. Additionally, we could take into account how often the user visited the dataset (view count) or just whether the dataset is "least recently used" (we could maintain a LRU list of datasets per user).
The view count would have the benefit of giving us a statistic which is potentially useful for other scenarios. However, we would need additionally tweaking to avoid that fresh datasets with a low view count "lose" against older datasets with a high view count.

Either way, the backend would probably need to register each "dataset view" (also triggered by opening a dataset).

### PR DESCRIPTION
**Pull request by fm3**
### URL of deployed dev instance (used for testing):
- https://lrudatasets.webknossos.xyz

### Steps to test:
- click on dataset
- after that, navigate to dataset list view, with browser network tab open
- in list reply, datasets should have `lastUsedByUser` field, non-zero for the one you clicked on before

- the sorting itself is disabled for now, to first gather some data, see comment below

### Issues:
- fixes #2917

------
- ~~[ ] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)~~
- [x] Updated [migration guide](../blob/master/MIGRATIONS.md#unreleased) if applicable
- [x] Ready for review

### CODE REVIEW

**Comment by fm3** [comment]
@philippotto @daniel-wer backend part is done, dataset json now has a `lastUsedByUser` timestamp (0 if never used) – does one of you want to create the frontend changes? Would null be a better default value than 0? Then i can change that.

**Comment by philippotto** [comment]
> @philippotto @daniel-wer backend part is done, dataset json now has a lastUsedByUser timestamp (0 if never used) – does one of you want to create the frontend changes? Would null be a better default value than 0? Then i can change that.

Awesome, I can implement the corresponding front-end changes. 0 should be fine.

**Comment by philippotto** [comment]
I added the necessary front-end code. The sorting is not enabled for now, though. I'd suggest to deploy this so that the LRU data is collected. A few days later, we can enable the sorting. Otherwise, it might be counterintuitive why the sorting only respects some recently used datasets but not all. What do you think, @fm3 @daniel-wer?

File: app/assets/javascripts/dashboard/advanced_dataset/advanced_dataset_view.js [hunk 1]
```diff
@@ -72,20 +73,35 @@ class AdvancedDatasetView extends React.PureComponent<Props, State> {
...
-            .map(row => ({
-              row,
-              diceCoefficient: dice(row.name, this.props.searchQuery),
-            }))
-            .sortBy("diceCoefficient")
-            .map(({ row }) => row)
+            .map(dataset => {
+              const diceCoefficient = dice(dataset.name, this.props.searchQuery);
+              const rank = useLruRank ? datasetToRankMap.get(dataset) || 0 : 0;
+              const rankCoefficient = rank / filteredDataSource.length;
```
Comment (daniel-wer) [inline app/assets/javascripts/dashboard/advanced_dataset/advanced_dataset_view.js]: As discussed, this should be `1 - (rank / filteredDataSource.length)` as the dice coefficient is 1 for two identical entities and, therefore, the rankCoefficient should also be 1 for the least recently used dataset.

**Pull request merged**